crates/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,452 @@
+
+#[derive(Debug, PartialEq)]
+enum AssignmentReferenceKind {
+    GlobalName(AtomIndex),
+    #[allow(dead_code)]
+    Prop(AtomIndex),
+    #[allow(dead_code)]
+    Elem,
+}
+
+// See AssignmentReferenceEmitter.
+// This uses struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct AssignmentReference {
+    kind: AssignmentReferenceKind,
+}
+impl AssignmentReference {
+    fn new(kind: AssignmentReferenceKind) -> Self {
+        Self { kind }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum CallKind {
+    Normal,
+    // FIXME: Support eval, Function#call, Function#apply etc.
+}
+
+// See *ReferenceEmitter.
+// This uses struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct CallReference {
+    kind: CallKind,
+}
+impl CallReference {
+    fn new(kind: CallKind) -> Self {
+        Self { kind }
+    }
+}
+
+// Struct for emitting bytecode for get `name` operation.
+pub struct GetNameEmitter<'alloc> {
+    pub name: &'alloc str,
+}
+impl<'alloc> GetNameEmitter<'alloc> {
+    pub fn emit(self, emitter: &mut AstEmitter) {
+        let name_index = emitter.emit.get_atom_index(self.name);
+
+        //              [stack]
+
+        // FIXME: Support non-global case.
+        emitter.emit.get_g_name(name_index);
+        //              [stack] VAL
+    }
+}
+
+// Struct for emitting bytecode for get `obj.key` operation.
+pub struct GetPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub obj: F,
+    pub key: &'alloc str,
+}
+impl<'alloc, F> GetPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        let key_index = emitter.emit.get_atom_index(self.key);
+
+        //              [stack]
+
+        (self.obj)(emitter)?;
+        //              [stack] OBJ
+
+        emitter.emit.get_prop(key_index);
+        //              [stack] VAL
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for get `super.key` operation.
+pub struct GetSuperPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub this: F,
+    pub key: &'alloc str,
+}
+impl<'alloc, F> GetSuperPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        let key_index = emitter.emit.get_atom_index(self.key);
+
+        //              [stack]
+
+        (self.this)(emitter)?;
+        //              [stack] THIS
+
+        emitter.emit.callee();
+        //              [stack] THIS CALLEE
+
+        emitter.emit.super_base();
+        //              [stack] THIS OBJ
+
+        emitter.emit.get_prop_super(key_index);
+        //              [stack] VAL
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for get `obj[key]` operation.
+pub struct GetElemEmitter<F1, F2>
```

Comment:
The main reason I am not excited about this patch is that it isn't clear why these should be structs.

GetElemEmitter seems no better than a single function:

```
pub fn emit_get_elem(emitter: &mut AstEmitter, obj: F1, key: F2) -> Result<(), EmitError>
where
    F1: ...
    F2: ...
{
    obj(emitter)?;
    key(emitter)?;
    emitter.emit.get_elem();
    Ok(())
}
```

This is less API surface (one `pub` feature instead of four).

Reply:
I'm using struct to clarify which parameter is for which AST node
(rust doesn't yet have named parameter)
given that lambda can have same type for different AST node, it's better naming them,
in the callsite.